common/transport/amqp/src/amqp.ts

```diff
@@ -136,59 +135,26 @@ export class Amqp {
             debug('received an error while disconnected: maybe a bug: ' + (!!err ? err.toString() : 'falsy error object.'));
           },
           amqpDisconnected: () => debug('ignoring disconnected event while disconnected'),
-          connect: (policyOverride, connectCallback) => {
-            this._fsm.transition('connecting', policyOverride, connectCallback);
+          connect: (config, connectCallback) => {
+            this._fsm.transition('connecting', config, connectCallback);
           },
           disconnect: (callback) => callback(null, new results.Disconnected()),
-          attachSenderLink: (endpoint, linkOptions, callback) => {
-            this._fsm.handle('connect', null, (err) => {
-              if (err) {
-                callback(err);
-              } else {
-                this._fsm.handle('attachSenderLink', endpoint, linkOptions, callback);
-              }
-            });
-          },
-          attachReceiverLink: (endpoint, linkOptions, callback) => {
-            this._fsm.handle('connect', null, (err) => {
-              if (err) {
-                callback(err);
-              } else {
-                this._fsm.handle('attachReceiverLink', endpoint, linkOptions, callback);
-              }
-            });
-          },
+          attachSenderLink: (endpoint, linkOptions, callback) => callback(new errors.NotConnectedError()),
```

Comment:
>callback(new errors.NotConnectedError()) [](start = 65, length = 40)

I'm really not sure how I feel about this. On one hand, I'm all for more simple code and symmetry between MQTT and AMQP. on the other hand I wonder if the right thing is to align AMQP with MQTT and not the other way around. 
These layers should be independent. we should trust each layer to do the right thing as much as possible. I'm not going to question this that much for now especially since Tony is coming up with big changes at this level to incorporate rhea, but this feels like a philosophical change that slips through as part of an API adjustment that really should not require it (it's "just" a user agent string).

Reply:
I agree that it feels strange to fit in a philosophical change like this, and I feel like I dropped this on you without discussing beforehand, and I apologize for that.  

For context, here’s some of my rationale:  
1. At this layer,  we don’t know who we’re connecting to until the caller calls connect().  So, for example, it doesn’t make sense to call attachSenderLink before calling connect because we wouldn't know the destination URI.  The only time this code would work correctly is when the connection is established, then somehow disconnected.  (the device-level transport gets the destination URI at construction time, so it does make sense at that level).
2. The layers above take care of connecting before doing any of these operations, so, at best, this functionality is redundant.  At worst, this layer would be trying to circumvent the policy of upper layers, in effect deciding _when_ to reconnect.  
3. I agree that this could increase robustness by “just doing the right thing”, but it could also cover up bugs.  Because upper layers are responsible for reconnecting, this code would only come into play when the connection is broken but upper layers don’t know that it’s broken.  This seems like something that should be addressed rather than covered up by lower layers.  In effect, this would be a lower layer covering up the bugs in an upper layer, and it’s usually the other way around.
4. Because there’s no known  case where upper layers would address this functionality, the only coverage we get is at the unit test level.  As far as we know, this code _never_ gets executed in a product setting. 
5. Having 2 different layers be responsible for reconnecting just smells bad.